### ISSUE TITLE
resizing token embeddings causes output embedding to be reinitialized in `post_init` when `tie_word_embedding` is False

### ISSUE THREAD

**Issue by avishaiElmakies** [huggingface/transformers#35141]
### System Info

- `transformers` version: 4.46.3
- Platform: Linux-6.6.20-aufs-1-x86_64-with-glibc2.36
- Python version: 3.11.2
- Huggingface_hub version: 0.26.1
- Safetensors version: 0.4.5
- Accelerate version: 1.0.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.5.1+cu124 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA A10

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

This code reproduces the problem:

```
pythia = AutoModelForCausalLM.from_pretrained("EleutherAI/pythia-410m")
pythia.resize_token_embeddings(502)
pythia.post_init()
```

the default value for `tie_word_embeddings` in pythia is False.
I believe the problem arises from the fact the if `tie_word_embeddings` is False, Then `resize_token_embeddings` creates a new `nn.Linear` object that doesn't have the flag `_is_hf_initialized`(causing it to be `False` when using `getattr`), and then `post_init` calls `_init_weights` on the new module.

https://github.com/huggingface/transformers/blob/c8c8dffbe45ebef0a8dba4a51024e5e5e498596b/src/transformers/modeling_utils.py#L2406
 

### Expected behavior

`post_init` should not change the weights of output_embeddings after a resize.

**Comment by Rocketknight1** [comment]
@avishaiElmakies this might be a bug, but can you explain why you're calling `post_init()` after the layer resize?

**Comment by avishaiElmakies** [comment]
I am using the model inside a more complex model which i add more modules to. I want thow to be initiated normally, which from what i understand should be done using post_init.

**Comment by Rocketknight1** [comment]
@avishaiElmakies I understand! In that case, would setting `_is_hf_initialized` in `resize_token_embeddings` fix the issue? Would you be willing to make that PR?

**Comment by avishaiElmakies** [comment]
@Rocketknight1 yes. I think the solution is `_is_hf_initialized` in the function would fix the issue. I would be willing to do a PR, but don't think I will have the time this week. so if someone else wants to do it will be fine

**Comment by Rocketknight1** [comment]
@avishaiElmakies thank you! The PR isn't urgent, but we'd definitely appreciate the fix if you get a chance.

**Comment by avishaiElmakies** [comment]
Thank you for the response!

**Comment by ArthurZucker** [comment]
PR is most welcome as I am not sure this is the intended API, but might be good to have!

**Comment by sambhavnoobcoder** [comment]
Hi @Rocketknight1 @ArthurZucker , 
I saw this open and tried to tackle it . For some reason , on the surface of it , it seems like a really simple issue , so i have raised a PR #36221  for it , however , considering the simplicity of the solution , i am a bit in doubt if my understanding of the issue is correct , so please look at the PR , and in case i have misunderstood the issue at its core , correct me to point me in the right direction.
